bootstrap/codegen_ssa: ship llvm-strip and use it for -Cstrip

davidtwco opened the pull request:
<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r\? <reviewer name> (with the `\` removed)
-->
<!-- homu-ignore:end -->
Fixes #131206.

- Includes `llvm-strip` (a symlink to `llvm-objcopy`) in the compiler dist artifact so that it can be used for `-Cstrip` instead of the system tooling.
- Uses `llvm-strip` instead of `/usr/bin/strip` for macOS. macOS needs a specific linker and the system one is preferred, hence #130781 but that doesn't work when cross-compiling, so use the `llvm-strip` utility instead.

cc #123151

davidtwco committed:
bootstrap: include `llvm-objcopy` in dist

davidtwco committed:
codegen_ssa: use `llvm-objcopy` for macOS strip

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11237903214/job/31241412784) [(plain)](https://github.com/rust-lang/rust/commit/159f31c1028d7ead09e8b3ebe6eb6c518d3eef6a/checks/31241412784/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 2.742 Building wheels for collected packages: reuse
#16 2.743   Building wheel for reuse (pyproject.toml): started
#16 2.993   Building wheel for reuse (pyproject.toml): finished with status 'done'
#16 2.994   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132715 sha256=dfa09868353292d98f811d3efdb0d54d07389e808efc71d68e3b93c514bf8bec
#16 2.994   Stored in directory: /tmp/pip-ephem-wheel-cache-bfdqgot0/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#16 2.997 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#16 3.391 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#16 3.392 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#16 DONE 3.5s
---
    Checking xz2 v0.1.7
error: the borrowed expression implements the required traits
    --> src/core/build_steps/compile.rs:1975:53
     |
1975 |             t!(builder.symlink_file("rust-objcopy", &libdir_bin.join(dst_exe)));
     |                                                     ^^^^^^^^^^^^^^^^^^^^^^^^^ help: change this to: `libdir_bin.join(dst_exe)`
     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_borrows_for_generic_args
     = note: `-D clippy::needless-borrows-for-generic-args` implied by `-D warnings`
     = help: to override `-D warnings` add `#[allow(clippy::needless_borrows_for_generic_args)]`

```

</details>

jieyouxu commented on src/bootstrap/src/core/build_steps/dist.rs:
```diff
@@ -473,6 +473,14 @@ impl Step for Rustc {
             }
+
+            {
+                let src_dir = builder.sysroot_libdir(compiler, host).parent().unwrap().join("bin");
```
Remark: this seems like it should be `sysroot_bindir` if it doesn't already exist

davidtwco replied:
Makes sense, fixed.

lqd commented on src/bootstrap/src/core/build_steps/compile.rs:
```diff
@@ -1965,6 +1965,16 @@ impl Step for Assemble {
+            let src_exe = exe("llvm-objcopy", target_compiler.host);
+            let dst_exe = exe("rust-objcopy", target_compiler.host);
+            builder.copy_link(&libdir_bin.join(src_exe), &libdir_bin.join(dst_exe));
+            let dst_exe = exe("rust-strip", target_compiler.host);
```
do you really need the symlink, since the strip command used in the compiler is `rust-objcopy`?

davidtwco replied:
Removed the symlink, this was a holdover from before I hit the "symlink can't be added to tarball" check.

lqd commented on src/bootstrap/src/core/builder.rs:
```diff
@@ -1151,6 +1151,11 @@ impl<'a> Builder<'a> {
         self.ensure(compile::Sysroot::new(compiler))
     }
 
+    /// Returns the bindir for a compiler's sysroot.
```
Probably t-bootstrap will have opinions here, because there are 2 "bin" directories 😓

workingjubilee commented:
Thank you for posting this. I tried twice and couldn't even figure out where to start.

taiki-e commented:
> illumos 

AFAIK some change is needed for illumos too because using the default strip binary of Linux host causes a problem (https://github.com/oxidecomputer/helios/issues/147).
However, I have not actually tested to see if using LLVM strip would solve the problem. ([This comment](https://github.com/rust-lang/rust/blob/68e4d9654ea5bd3553c892a23f42b4f17b37e5f1/compiler/rustc_codegen_ssa/src/back/link.rs#L1107-L1109) may indicate that anything other than native strip utility would not work.)

cc @sunshowers

jclulow commented:
> @jclulow (we have no target maintainers for the illumos targets in the documentation, but I've seen @jclulow in the issue tracker and I know is involved in illumos)

Hey, with my apologies, someone pointed this out to me recently and I owe a PR to fix the docs and list some maintainers there.

> We could make this same change to the illumos and AIX targets which also hardcode a path to a strip binary. Pinging relevant people for those targets to see if this would be desirable.

When we're running on an illumos host and compiling for an illumos target, we definitely want to use the system **strip** binary explicitly.  If we're cross compiling, either on an illumos host for a non-illumos target, or on a non-illumos host, I don't have a particular preference and we should do whatever is tested to work there.

Does that make sense?

bzEq commented:
On AIX, we are still using `/usr/bin/strip`. `llvm-strip` is not compatible with AIX's system `strip` utility.

bors commented:
:umbrella: The latest upstream changes (presumably #131495) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

davidtwco commented:
> Thanks for headsup. On AIX, we are still using `/usr/bin/strip`. `llvm-strip` is not compatible with AIX's system `strip` utility.

Okay, I've kept AIX's logic as it is, but I have added a warning when cross-compiling that whatever is in `/usr/bin/strip` may not work because it isn't AIX's system `strip`.

> When we're running on an illumos host and compiling for an illumos target, we definitely want to use the system **strip** binary explicitly. If we're cross compiling, either on an illumos host for a non-illumos target, or on a non-illumos host, I don't have a particular preference and we should do whatever is tested to work there.
> 
> Does that make sense?

That makes sense. I've kept `/usr/bin/strip` if not cross-compiling, and we'll now use `llvm-strip` instead of `/usr/bin/strip` when cross-compiling (that seems more likely to work reliably than relying on whatever `/usr/bin/strip` could be). I also changed the condition we used here from `target.os == "illumos"` to `target.is_like_solaris`, since that's why we have `is_like_solaris` in the first place. Let me know if that's not correct and there is some difference with stripping between illumos and solaris in general.

davidtwco commented on compiler/rustc_codegen_ssa/src/back/link.rs:
```diff
@@ -1147,6 +1148,13 @@ fn strip_symbols_with_external_utility(
         cmd.arg(option);
     }
+
+    let mut new_path = sess.get_tools_search_paths(false);
+    if let Some(path) = env::var_os("PATH") {
+        new_path.extend(env::split_paths(&path));
+    }
+    cmd.env("PATH", env::join_paths(new_path).unwrap());
```
It probably is but we override the `PATH` for `rust-lld` so I'll keep it this way and we can change both together later.

lqd replied:
Ah, do we? I remember we set the linker search path to the lld wrappers in the sysroot without overriding the PATH (so they should execute `../rust-lld` without overrides), and that rustup ensures the PATH contains the bin and rustlib/bin folders, but didn’t remember rustc overrode it for rust-lld.

davidtwco replied:
https://github.com/rust-lang/rust/blob/bca5fdebe0e539d123f33df5f2149d5976392e76/compiler/rustc_codegen_ssa/src/back/linker.rs#L130

I believe this is where it happens

lqd replied:
Yeah that's indeed likely it for the targets involving the linker directly. We don't need/make use of that on the common linux/mac targets.

albertlarsan68 reviewed (commented):
LGTM for the technical side in bootstrap

albertlarsan68 commented:
r? t-compiler for the compiler side of things

bors commented:
:umbrella: The latest upstream changes (presumably #131988) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

cjgillot commented:
r? compiler

bors commented:
:umbrella: The latest upstream changes (presumably #132497) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

jieyouxu commented:
r? jieyouxu

jieyouxu commented on src/bootstrap/src/core/builder/mod.rs:
```diff
@@ -1161,6 +1161,11 @@ impl<'a> Builder<'a> {
         self.ensure(compile::Sysroot::new(compiler))
     }
 
+    /// Returns the bindir for a compiler's sysroot.
+    pub fn sysroot_bindir(&self, compiler: Compiler, target: TargetSelection) -> PathBuf {
+        self.sysroot_libdir(compiler, target).parent().unwrap().join("bin")
+    }
```
Remark (for myself): send a follow-up to avoid this dance

jieyouxu replied:
Follow-up PR: https://github.com/rust-lang/rust/pull/132723

jieyouxu reviewed (approved):
Thanks, this looks good to me (with compiler hat 🎩 on). Feel free to r=albertlarsan68 and me after a rebase.

davidtwco commented:
@bors=jieyouxu,albertlarsan68

davidtwco commented:
@bors r=jieyouxu,albertlarsan68

bors commented:
:pushpin: Commit f745467cd99cdcb62151c3b0c90f5467e7adfe69 has been approved by `jieyouxu,albertlarsan68`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=jieyouxu,albertlarsan68 f745467cd99cdcb62151c3b0c90f5467e7adfe69 -->
<!-- homu: {"type":"Approved","sha":"f745467cd99cdcb62151c3b0c90f5467e7adfe69","approver":"jieyouxu,albertlarsan68","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bjorn3 commented on src/bootstrap/src/core/build_steps/compile.rs:
```diff
@@ -1977,6 +1976,14 @@ impl Step for Assemble {
+            let src_exe = exe("llvm-objcopy", target_compiler.host);
+            let dst_exe = exe("rust-objcopy", target_compiler.host);
+            builder.copy_link(&libdir_bin.join(src_exe), &libdir_bin.join(dst_exe));
+        }
```
This breaks bootstrapping when the bootstrap compiler doesn't contain any LLVM tools as is the case when cg_clif bootstraps rustc:

```
Building stage0 tool lld-wrapper (x86_64-unknown-linux-gnu)
     Compiling lld-wrapper v0.1.0 (/home/runner/work/rustc_codegen_cranelift/rustc_codegen_cranelift/rust/src/tools/lld-wrapper)
      Finished `release` profile [optimized] target(s) in 0.43s
  thread 'main' panicked at src/lib.rs:1714:24:
src.symlink_metadata() failed with No such file or directory (os error 2) ("src = /home/runner/work/rustc_codegen_cranelift/rustc_codegen_cranelift/rust/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/bin/llvm-objcopy")
```

It should probably check `builder.config.llvm_enabled(target_compiler.host) && builder.config.llvm_tools_enabled` before copying to ensure that `llvm-objcopy` exists in the first place.

jieyouxu replied:
Thanks for the heads-up, I'll send a follow-up to fix that.

jieyouxu replied:
Follow-up: https://github.com/rust-lang/rust/pull/132720
